### PR TITLE
[ST-0025] Tag-based test execution filtering

### PR DESCRIPTION
**Pull request by gmedori**
For convenience, here's a link to the implementation of the feature described in this proposal: https://github.com/swiftlang/swift-testing/pull/1531

### CODE REVIEW

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+## Integration with Supporting Tools
+
+If a codebase has test functions or suites that contain the string `tag:`, then any filtering/skipping arguments that begin with `tag:` behave differently with this proposal. They are treated as tags rather than regular expressions.
+
+It's worth noting that string `tag:` can only appear in symbol names when using raw identifiers as described above. I feel that this scenario is relatively rare and that this improvement is worth the edge case.
+
+## Future Directions
+
+N/A
```
Comment (grynspan) [inline proposals/testing/nnnn-tag-based-test-execution-filtering.md]: Consider whether there are other aspects of tests on which devs may want to filter?

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ 
+ As of [SE-0451](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0451-escaped-identifiers.md), Swift has raw identifiers which means the following is valid Swift:
+
+```swift
+@Test func `tag:uiTest`() { /* ... */ }
+```
+
+To continue to allow matching for such function names, we will also allow the colon character to be escaped such that the entire argument will be treated as a single regular expression:
+
+```
```
Comment (grynspan) [inline proposals/testing/nnnn-tag-based-test-execution-filtering.md:74]:
```suggestion
```sh
```

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+swift test --skip 'tag\:uiTest'
+```
+
+The example above would behave as though the string `tag:uiTest` were passed as a regular expression, omitting the escaping backslash in the final regular expression.
+
+> **Note**: Most shells treat the backslash character `\` as a special character used for escaping. In order for the application to receive it, the argument needs to either be wrapped in quotes like `'tag\:uiTest'`, or the backslash itself needs to be escaped, `tag\\:uiTest`.
```
Comment (grynspan) [inline proposals/testing/nnnn-tag-based-test-execution-filtering.md]:
```suggestion
> [!NOTE]
> Most shells treat the backslash character `\` as a special character used for escaping. In order for the application to receive it, the argument needs to either be wrapped in quotes like `'tag\:uiTest'`, or the backslash itself needs to be escaped, `tag\\:uiTest`.
```

See https://github.com/orgs/community/discussions/16925

Comment (grynspan) [inline proposals/testing/nnnn-tag-based-test-execution-filtering.md]: Should we just make it a policy that `\` escapes the next character? If not, consider adding some discussion in this area to **Alternatives Considered**.

Or, if that's impractical or undesirable, should we document how to write `\:` verbatim (i.e. we probably also need to allow escaping `\` itself)?

File: proposals/testing/0025-tag-based-test-execution-filtering.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+> **Note**: Most shells treat the backslash character `\` as a special character used for escaping. In order for the application to receive it, the argument needs to either be wrapped in quotes like `'tag\:uiTest'`, or the backslash itself needs to be escaped, `tag\\:uiTest`.
+
+## Source Compatibility
+
+The change is an implementation detail and makes no changes to source compatibility.
+
+## Integration with Supporting Tools
+
+If a codebase has test functions or suites that contain the string `tag:`, then any filtering/skipping arguments that begin with `tag:` behave differently with this proposal. They are treated as tags rather than regular expressions.
```
Comment (grynspan) [inline proposals/testing/0025-tag-based-test-execution-filtering.md:111]: This discussion might belong in **Source Compatibility** rather than this section because it doesn't really affect e.g. VS Code or Xcode integrations.

File: proposals/testing/0025-tag-based-test-execution-filtering.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+```
+swift test --skip tag:uiTest --skip tag:integrationTest
+```
+
+It is not currently possible to create a filter/skip based on _all_ tags specified, as opposed to _any_ of the tags specified. In other words, supplying multiple `--filter`/`--skip` options is an **or** operation, not an **and** operation.
+
+### Handling Raw Identifiers
+ 
+ As of [SE-0451](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0451-escaped-identifiers.md), Swift has raw identifiers which means the following is valid Swift:
```
Comment (allevato) [inline proposals/testing/0025-tag-based-test-execution-filtering.md]: You should also consider what needs to happen (if anything different) if the tag itself is a raw identifier (unless anything else in Swift Testing prohibits that, which I don't know off the top of my head):

```swift
@Test(.tags(.`some tag with spaces`))
func `my test`() { ... }
```

I think I would expect to write `swift test --skip 'tag:some tag with spaces'` here.

Comment (grynspan) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: Then there's an open question about whether or not developers should specify backticks (which also affect shell processing).

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md (hunk 1 above)
Comment (gmedori) [reply proposals/testing/nnnn-tag-based-test-execution-filtering.md]: Ah fair point! I actually wrote this a little while ago to address an open github issue and am just putting up the PR now, and I didn't stop to think about future directions. Lemme give it a little thought.

File: proposals/testing/0025-tag-based-test-execution-filtering.md (hunk 4 above)
Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md:111]: Yeah I read through that last night and think I agree.

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md (hunk 3 above)
Comment (gmedori) [reply proposals/testing/nnnn-tag-based-test-execution-filtering.md]: _Personally_ I think that if we want to make that a policy, the policy should exist at SPM level for every command. I'm not sure it's useful to have a policy that only applies to swift testing. Probably skews impractical IMO. Think this is something we want to pursue anyway?

> Or, if that's impractical or undesirable, should we document how to write \: verbatim (i.e. we probably also need to allow escaping \ itself)?

Do you mean in the usage description of the `--skip` and `--filter` options?

File: proposals/testing/0025-tag-based-test-execution-filtering.md (hunk 5 above)
Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: ha, yep that complicates things. I think the most correct answer here is to write it as @allevato put it. Involving backticks opens the command up to different interpretations in different shells, where as single-quoting is relatively universal, even going back to shells like `sh` and `csh`.

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md (hunk 1 above)
Comment (gmedori) [reply proposals/testing/nnnn-tag-based-test-execution-filtering.md]: Added some thoughts in 61e2f4b2

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md (hunk 3 above)
Comment (gmedori) [reply proposals/testing/nnnn-tag-based-test-execution-filtering.md]: Resolved in d80d77b7

File: proposals/testing/nnnn-tag-based-test-execution-filtering.md (hunk 2 above)
Comment (gmedori) [reply proposals/testing/nnnn-tag-based-test-execution-filtering.md:74]: Resolved in d80d77b7

File: proposals/testing/0025-tag-based-test-execution-filtering.md (hunk 5 above)
Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: Added some more discussion around this scenario in 90054874

File: proposals/testing/0025-tag-based-test-execution-filtering.md (hunk 4 above)
Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md:111]: I moved this discussion to **Source Compatibility** and added some clearer thoughts around tool integrations in 90054874

File: proposals/testing/0025-tag-based-test-execution-filtering.md (hunk 5 above)
Comment (allevato) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: As a parallel, when using raw identifiers in the `-module-alias` or `-module-abi-name` command line flags of the compiler, you _don't_ need to include the backticks: you'd write `-module-abi-name 'foo bar'`, not ``-module-abi-name '`foo bar`'``. So I think it's reasonable to have the same policy here.

In other words, the backticks are required to delimit the identifier in _source code_ (and they also appear in the mangling of the symbol for compatibility reasons), but it's reasonable to drop them elsewhere.

Comment (grynspan) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: I don't have an opinion either way, but we should be clear about it. (And if the developer _does_ type the backticks, does that make it _wrong_?)

Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: > if the developer does type the backticks, does that make it _wrong_?

I could go either way on this, so IMO we should just fall back to precedent. From what @allevato said:

> you'd write `-module-abi-name 'foo bar'`, not ```-module-abi-name '`foo bar`'```

it _sounds_ like using backticks would be incorrect. So I'd opt to not do any special handling around stripping them for the user. But please correct me if I'm misinterpreting.

Comment (allevato) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: If you include the backticks in `-module-abi-name`, it looks like it checks whether the symbol needs to be escaped (which it does, because it contains a space) and then inserts a second pair of backticks, so the mangled symbol ends up being ` ``foo bar`` `. This is actually wrong because raw identifiers are not supposed to be able to contain backticks (syntactically, at least). The frontend should prevent that instead of letting it through.

So maybe that flag wasn't the best one to cite, but my vote here would be "don't require backticks and fail if the user provides them".

Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: I think this might be a common enough occurrence that there's actually value in detecting whether backticks were supplied and presenting a targeted error message. I'll work on adding that to the implementation later today.

Comment (gmedori) [reply proposals/testing/0025-tag-based-test-execution-filtering.md]: I was able to sort this out! Apologies for the delay. I've update the proposal with my suggestion here: a2b375a4

tl;dr: we'll alert the user that the symbol name shouldn't have backticks surrounding and strip them transparently. So it'll match what the user probably intended.
